Add a Physics-using component to current components

GiudGiud opened the pull request:
refs #24103

cutting up the old large PR for physics still.
we could delete the old FlowComponentNS or keep both, both are simply transition options, with this one being more general but having less-convenient syntax in a way

GiudGiud committed:
Apply parameter is useless if it does not override the default

GiudGiud committed:
Add a component to handle a physics definition

GiudGiud committed:
Add a test for adding a single physics to a single component

GiudGiud committed:
Add ability for components to query AD functors

GiudGiud committed:
Update to new syntax for physics component tests
Enable valgrind

GiudGiud committed:
Adapting input file for merging Physics and new syntax

GiudGiud committed:
Update syntax
Fix test definition
Add documentation to new component
refs #23794 #24103

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2254090/) on aa277dd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4409da00956c2511d3f6c6db886956a265cd4dd9`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2257755/) on 387714f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 843ddc9ad7abe9002a0444fb8bf3dce129299805`

GiudGiud committed:
Default physics to being defined everywhere

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2258308/) on fd75947 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27729/site)

This comment will be updated on new commits.

GiudGiud committed:
Relax dependency on problem for NSFVAction processMesh by moving a call to addNSVariables
Re-organize task dependencies to allow physics being initialized after componentse

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2272838/) on 13732f2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0445d490c50428b8ce3cbf4f27e1b017456306ff`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2272885/) on 8d5371e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27729/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0445d490c50428b8ce3cbf4f27e1b017456306ff`

GiudGiud committed:
Remove retrieval of FPs at construction in fluid property functions
- lets us construct functions much earlier. As FPs depend on variables

GiudGiud commented:
I ll be able to fix NS but if the Griffin thing is related that's going to be difficult

GiudGiud committed:
Move displaced problem init earlier to have functions functors
Add TI to displaced problem later as it is not ready when problem is created

GiudGiud committed:
Make sure displaced problem gets destructed earlier

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2278105/) on 4ef7830 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ca913d</th>
      <th colspan=4>#27729 4ef783</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.08%</td>
      <td>85.08%</td>
      <td>+0.00%</td>
      <td>93.75%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104117</td>
      <td>104127</td>
      <td>+10</td>
      <td>15</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18265</td>
      <td>18264</td>
      <td>-1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>ca913d</th>
      <th colspan=4>#27729 4ef783</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.48%</td>
      <td>+0.01%</td>
      <td>90.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7261</td>
      <td>7266</td>
      <td>+5</td>
      <td>18</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1234</td>
      <td>1234</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage/fluid_properties)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>ca913d</th>
      <th colspan=4>#27729 4ef783</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.18%</td>
      <td>84.17%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17037</td>
      <td>17031</td>
      <td>-6</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3201</td>
      <td>3203</td>
      <td>+2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage/navier_stokes)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>ca913d</th>
      <th colspan=4>#27729 4ef783</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.16%</td>
      <td>89.14%</td>
      <td>-0.02%</td>
      <td>80.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>14169</td>
      <td>14191</td>
      <td>+22</td>
      <td>24</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1723</td>
      <td>1729</td>
      <td>+6</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27729/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27729/coverage/xfem)

</details>

## Warnings

- `thermal_hydraulics` new line coverage rate 80.00% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
Ready for review imo

I raised https://github.com/libMesh/libmesh/issues/3882 for the destruction ordering issue that this PR avoids in a dirty-ish way

joshuahansel commented on modules/thermal_hydraulics/test/tests/junctions/wcnsfv/two_components_same_vars.i:
```diff
@@ -0,0 +1,140 @@
+# Operating conditions
```
Is there no test for this input file?

GiudGiud replied:
lemme check

GiudGiud replied:
deleting it

GiudGiud replied:
looking at a griffin issue then I ll reupload

joshuahansel commented on modules/thermal_hydraulics/test/tests/junctions/wcnsfv/two_components_same_vars.i:
```diff
@@ -0,0 +1,140 @@
+    energy_wall_function = '1'
+
+    outlet_boundaries = 'comp1:right'
+    # momentum_outlet_types = 'fixed-pressure'
+    # pressure_function = 'f1'
```
```suggestion
```

joshuahansel reviewed (changes requested):
Also, were you wanting to address the libMesh issue before this to avoid "dirtiness"?

GiudGiud commented:
> Also, were you wanting to address the libMesh issue before this to avoid "dirtiness"?

depends on what we decide to do.

joshuahansel commented:
> depends on what we decide to do.

I say if there's no urgency for this PR, have that conversation on the libmesh issue and make a decision on what's done there before proceeding with this PR.

GiudGiud commented:
no urgency

I just feel better when I have like 1 or so PR to work on rather than 5 :D

joshuahansel commented on framework/include/executioners/SolveObject.h:
```diff
@@ -41,7 +41,9 @@ class SolveObject : public MooseObject, public PerfGraphInterface, public Postpr
-  std::shared_ptr<DisplacedProblem> _displaced_problem;
+  /// We would rather have a shared pointer here. But the displaced problem currently must be destructed
+  /// before the displaced mesh, and keeping a shared pointer here delays this
+  DisplacedProblem * _displaced_problem;
```
So I guess only this move to raw pointer is the "dirtiness".

GiudGiud replied:
it s kind of being endorsed in that discussion actually.
Alex says use less shared_ptr

joshuahansel replied:
Oh sorry I misunderstood, I thought "my current solution" meant the changes you were proposing to make in libmesh, which would avoid the changes made here.

joshuahansel reviewed (changes requested):
Waiting for libmesh update to avoid switching to raw pointer. Otherwise looks good.

joshuahansel reviewed (approved):
I misunderstood the plan that was discussed. The move to raw pointer is preferred.

lindsayad commented on framework/include/executioners/SolveObject.h:
```diff
@@ -41,7 +41,9 @@ class SolveObject : public MooseObject, public PerfGraphInterface, public Postpr
   FEProblemBase & _problem;
   /// Displaced problem
-  std::shared_ptr<DisplacedProblem> _displaced_problem;
+  /// We would rather have a shared pointer here. But the displaced problem currently must be destructed
+  /// before the displaced mesh, and keeping a shared pointer here delays this
```
I don't know that we would rather have a shared pointer here. I don't see any reason why any `SolveObject` (there could be many of them) should be considered an owner of the displaced problem

GiudGiud replied:
yeah this is outdated. I ll delete the comment

GiudGiud commented:
Thanks for the review and thanks for the support! I feel like I'm walking the plank sometimes on components & physics
